src/azul/service/responseobjects/storage_service.py

```diff
@@ -39,12 +40,135 @@ def put(self, object_key: str, data: bytes, content_type: Optional[str] = None)
 
     def get_presigned_url(self, key: str) -> str:
         return self.client.generate_presigned_url(ClientMethod='get_object',
-                                                  Params=dict(Bucket=self.__bucket_name, Key=key))
+                                                  Params=dict(Bucket=self.bucket_name, Key=key))
 
     def create_bucket(self, bucket_name: str = None):
-        self.client.create_bucket(Bucket=(bucket_name or self.__bucket_name))
-        logger.warning(f'{type(self).__name__}: Created a bucket called "{bucket_name or self.__bucket_name}"')
+        self.client.create_bucket(Bucket=(bucket_name or self.bucket_name))
+
+
+class MultipartUploadHandler:
+    """
+    S3 Multipart Upload Handler
+
+    This class is to facilitate multipart upload to S3 storage. The class itself
+    is also a context manager.
+
+    Here is the sample usage.
+
+    .. code-block:: python
+
+       with MultipartUploadHandler('samples.txt', 'text/plain'):
+           handler.push(b'abc')
+           handler.push(b'defg')
+           # ...
+
+    where the context pattern will automatically shutdown upon exiting the context.
+    """
+
+    def __init__(self, object_key, content_type: str):
+        self.bucket_name = config.s3_bucket
+        self.object_key = object_key
+        self.upload_id = None
+        self.mp_upload = None
+        self.next_part_number = 1
+        self.content_type = content_type
+        self.parts = []
+        self.futures = []
+        self.thread_pool = None
+
+    def __enter__(self):
+        api_response = boto3.client('s3').create_multipart_upload(Bucket=self.bucket_name,
+                                                                  Key=self.object_key,
+                                                                  ContentType=self.content_type)
+        self.upload_id = api_response['UploadId']
+        self.mp_upload = boto3.resource('s3').MultipartUpload(self.bucket_name, self.object_key, self.upload_id)
+        self.thread_pool = ThreadPoolExecutor(max_workers=MULTIPART_UPLOAD_MAX_WORKERS)
+        return self
+
+    def __exit__(self, etype, value, traceback):
+        if etype:
+            logger.error('Upload %s: Error detected within the MPU context.\n\n%s',
+                         self.upload_id,
+                         '\n'.join(format_exception(etype, value, traceback)))
+            self.__abort()
+            return
+        self.__complete()
+
+    def __complete(self):
+        if not self.parts:
+            self.__abort()
+            raise EmptyMultipartUploadError(f'{self.bucket_name}/{self.object_key}')
```

Comment:
Originally, I want to prevent empty object from being uploaded. This is removed per your recommendation until there is such a use case that we have to handle.